Add search page module and remove controller extension

sascha-karnatz opened the pull request:
The controller methods were available on every page in Alchemy and was listening to the query param. This approach is not really extendable, if you need multiple parameter to process the search result. The main logic was moved into a SearchPage module which has nearly the same structure as before, but is now way more flexible, because it is only used in this particular page layout. It also allows the application to extend the search and the view without monkey patching the gem. The only downside is, that it isn't possible anymore to add a search string the searchresults - element to test the search in the admin interface.

sascha-karnatz committed:
Add search page module and remove controller extension

The controller methods were available on every page in Alchemy and was listening to the query param. This approach is not really extendable, if you need multiple parameter to process the search result.
The main logic was moved into a SearchPage module which has nearly the same structure as before, but is now way more flexible, because it is only used in this particular page layout. It also allows the application to extend the search and the view without monkey patching the gem.
The only downside is, that it isn't possible anymore to add a search string the searchresults - element to test the search in the admin interface.

tvdeyen commented on spec/dummy/app/views/alchemy/page_layouts/_search.html.erb:
```diff
@@ -1,2 +1,3 @@
+<% @search_results = Alchemy::Search::SearchPage.perform_search(params, ability: current_ability) %>
 <%= render_search_form %>
-<%= render_elements %>
+<%= render_search_results %>
```
```suggestion
<%= render "alchemy/search/results", search_results: search_results %>
```

tvdeyen replied:
And adjust the `alchemy/search/results` partial so it loads the results from the local vars instead of the ivar

tvdeyen replied:
and remove the `render_search_results` helper

tvdeyen commented on spec/dummy/app/views/alchemy/page_layouts/_search.html.erb:
```diff
@@ -1,2 +1,3 @@
+<% @search_results = Alchemy::Search::SearchPage.perform_search(params, ability: current_ability) %>
```
```suggestion
<% search_results = Alchemy::Search::SearchPage.perform_search(params, ability: current_ability) %>
```

tvdeyen commented on README.md:
````diff
@@ -143,40 +146,12 @@ page layout as `searchresults: true`. The search form will pick this page as res
 
 ```erb
 <!-- app/views/alchemy/page_layouts/_search.html.erb -->
-<%= render_elements %>
````
We loose the ability to freely place the search results on the page with this change. I would like to keep the element

tvdeyen commented on README.md:
```diff
@@ -143,40 +146,12 @@ page layout as `searchresults: true`. The search form will pick this page as res
-  <%= render_search_results %>
-<%- end -%>
+<% search_results = Alchemy::Search::Page.perform_search(params, ability: current_ability) %>
+<%= render "alchemy/search/results", search_results: search_results %>
```
```erb
<!-- app/views/alchemy/elements/_searchresults.html.erb -->
<%= element_view_for(searchresults) do |el| -%>
  <% search_results = Alchemy::Search::Page.perform_search(params, ability: current_ability) %>
  <%= render "alchemy/search/results", search_results: search_results %>
<%- end -%>
```

sascha-karnatz committed:
Remove render_search_results - helper

The helper method was only a short cut for the results partial, but it was using the global search_results variable from the previous controller_methods extension. With the SearchPage module it isn't necessary anymore to have a global variable.
